valhalla/baldr/edgeinfo.h

```diff
@@ -290,12 +297,12 @@ class EdgeInfo {
     uint32_t speed_limit_ : 8;     // Speed limit (kph)
     uint32_t extended_wayid0_ : 8; // Next byte of the way id
 
-    uint32_t name_count_ : 4;          // How many name infos we expect
-    uint32_t encoded_shape_size_ : 16; // How many bytes long the encoded shape is
-    uint32_t extended_wayid1_ : 8;     // Next next byte of the way id
-    uint32_t extended_wayid_size_ : 2; // How many more bytes the way id is stored in
-    uint32_t has_elevation_ : 1;       // Does the edgeinfo have elevation?
-    uint32_t spare0_ : 1;              // not used
+    uint32_t name_count_ : 4;                  // How many name infos we expect
+    uint32_t encoded_shape_size_ : 16;         // How many bytes long the encoded shape is
+    uint32_t extended_wayid1_ : 8;             // Next next byte of the way id
+    uint32_t extended_wayid_size_ : 2;         // How many more bytes the way id is stored in
+    uint32_t has_elevation_ : 1;               // Does the edgeinfo have elevation?
+    uint32_t had_conditional_speed_limits : 1; // Does the edgeifo have conditional speed limits?
```

Comment:
I'm really not sure about consuming the last bit of the `EdgeInfo` header for `had_conditional_speed_limits` because of the backward-forward compatibility reasons.
Alternatively, I use this bit as a marker for the "second header" and store required info there.

Reply:
```cpp
mutable std::multimap<TaggedValue, std::string> tag_cache_;
```
Well, the only thing I don't like is putting u64 values into a string. Well, technically it is not a problem to put bytes and then do `std::reinterpret_cast<uint64_t*>(tag.second.data())`, it just looks... well, it looks strange.

Reply:
Alternatively, I can remove all the stuff around parsing, encoding and so on, and just keep conditional speed limits as an original string inside the tagged values (similar to how Valhalla handles linguistics, ), that Valhalla will simply return. @kevinkreiser wdyt about this approach?

Reply:
true, that would be the simplest option! i might still suggest you normalize the values by sending it through the parsing and to_string methods you have. on the flip side of that it means that we cant store every single strange permutation people put in the data (but that might be a good thing anyway?)

i guess id also note that its likely a good deal cheaper (maybe half?) to store the fixed size `DateRange` struct than the strings, which does very much appeal to me. in the end though i dont feel too strongly about either approach. id say i have a slight preference for the above, keep the parse and to_string stuff, round trip the raw values to normalize them and store the string in the key value. whats your preference?

Reply:
I think an upside of the normalization would be that it tells the user what sense valhalla actually made of a condition. E.g. if there's a "PH" value in that string (which means the speed only applies on public holidays, for which there's no support at the moment), and that string just goes straight into the `/locate` output I see people could be more likely to complain why it's not being applied on christmas eve :smile:

Reply:
> whats your preference?

I would like to leave all the of parsing the suprisingly complex [opening hours](https://wiki.openstreetmap.org/wiki/Key:opening_hours) format at the Valhalla side and return somehow structured output, so downstream users can write logic like `condition.dow_mask & (1 << today.get_weekday())`.

Another argument for keeping fixed size parsed value is that it makes possible to use it in the costing (or where else speeds are used) one day.

Reply:
> possible to use it in the costing 

this was the argument i was going to make but decided to not complicate the discussion, im glad you were thinking it too :smile:

Reply:
Talking about the costing - I can't find any `EdgeInfo` usage in the costing, so maybe some another place is more suitable for storing conditional speed limits?

Reply:
this is why i didnt want to complicate the discussion :smile: ill try to do so succinctly.

we purposely dont give edgeinfo to the costing because its expensive to pull out of the tile relative to the other items. its meant for things that are sparsely populated, conditional, or just vanity metrics not something we cost on in the tight loop of the algorithm.

that said you can get it because you get the tile and the edge. if you have the edge and you have the tile then you have a way to get the edgeinfo.

at the moment speed limit isnt that interesting to costing because its usually baked into the `directededge->speed` but in this case, if the algorithm was trackign time and the edge had conditional speed limits and the speed calculation needed to fall back to the speed limit (because no real time, no historical, etc), then it would make sense to pull out the conditions and check if one overlaps this time. as you can see it would be quite a complication (a rare one) but indeed, as you stated it would be better to have it in the parsed struct if we plan on doing that.

so how about we keep the fixed size struct, we throw it in with the tagged values and we save using it in GetSpeed for a rainy day :wink:

Reply:
Thanks for the great explanation! I like the approach you suggested, I just need to convince myself that it is a good idea to do such hack:
```cpp
// somewhere at writing
ConditionalSpeedLimit limit;
std::string value(reinterpret_cast<char*>(&limit), sizeof(limit));

// and then at reading
const std::string limit_str = edge_info.GetTags()[kConditionalSpeedLimit];
const uint8_t count = limit_str.front();
for (size_t pos = 1; pos < 1 + count * sizeof(ConditionalSpeedLimit); pos += sizeof(ConditionalSpeedLimit)) {
  const ConditionalSpeedLimit limit = *reinterpret_cast<ConditionalSpeedLimit*>(&limit_str.data() + pos);
  // use/check limit...
}
```

Reply:
All names (tagged values) are stored using null terminator, which should be hacked around as zero bytes are possible in the `ConditionalSpeedLimit`.

Reply:
As far as I got, it is not a problem to have strings with `\0` somewhere inside, for example we don't dump all tagged values into a json. So let's try - e79f54d39a65720e2432c155e8f1a9f4ce5e9f5c